pathscale/man/linux/man1/assign.1
.\" '\" Copyright (C) 2007 PathScale, LLC.  All Rights Reserved.
.\" '\"
.\" '\" Copyright (C) 2006, 2007 QLogic Corporation. All Rights Reserved.
.\" '\"
.\" '\" Copyright (C) 2003, 2004, 2005 PathScale, Inc.  All Rights Reserved.
.\" '\"
.\" '\"Contact information:  PathScale, LLC., 2071 Stierlin Court, Suite 200,
.\" '\"Mountain View CA 94043, USA, or:
.\" '\"
.\" '\"http://www.pathscale.com
.\" '\"
.TH "assign" "1" "" "PathScale, LLC." "PathScale Compiler Suite"
.SH "NAME"
\fBassign\fR
\- Assign customizations to Fortran file I/O
.SH "SYNOPSIS"
\fBassign\fR
[\fIcommand-options\fR]
[\fIcustomize-options\fR]
[\fIfilespec\fR]
.SH "DESCRIPTION"
The \fBassign\fR command customizes the behavior of Fortran I/O for a
certain set of files or logical units (for example, it can swap bytes
when using big-endian files on a little-endian computer). By issuing
the command several times, you can dictate different customizations
for each of several sets of files or units.
.PP
The \fIcommand-options\fR relate to the command itself; the
\fIcustomize-options\fR describe the desired behavior, and the
\fIfilespec\fR describes which files or logical units those
customizations apply to. Each command allows multiple
\fIcustomize-options\fR, but no more than one \fIcommand-option\fR and
one \fIfilespec\fR.
.PP
The command works by storing information into a file that the Fortran
library inspects before performing I/O. You must set the environment variable
FILENV to the name of that file before you type the \fBassign\fR
command, and it must remain set when the program runs.
.PP
The \fBpathf95\fR(1) compiler also provides compile-time options for
swapping bytes that are simpler to use. Setting FILENV overrides the
compiled-in choices.
.SH "FILESPEC"
The \fIfilespec\fR can be:
.TP
\fBu:unit\fR
The logical unit numbered \fIunit\fR.
.TP
\fBf:filename\fR
The file whose name is \fIfilename\fR.
.TP
\fBp:pattern\fR
Every file whose name matches \fIpattern\fR, using
the \fB%\fR symbol as a wildcard which matches zero or more characters.
.TP
\fBg:du\fR
Every direct unformatted file (e.g., files opened using "access='direct'
form='unformatted'").
.TP
\fBg:su\fR
Every sequential unformatted file (e.g., files opened using "access='sequential'
form='unformatted'").
.TP
\fBg:all\fR
Every file.
.SH "COMMAND OPTIONS"
.TP
\fB-R\fR
Remove all customizations for the specified \fIfilespec\fR.  If no
\fIfilespec\fR is specified, remove all customizations.
.TP
\fB-I\fR
Add this command to any previous ones.  Normally each \fBassign\fR
command erases the effect of all previous \fBassign\fR commands, even
if they refer to a different set of files.  To allow more than one
command to take effect at the same time, all but the first command
must use \fB\-I\fR.
.TP
\fB-O\fR
Erase the effect of previous \fBassign\fR commands before adding the
current command. This is the default.
.TP
\fB-V\fR
Display all \fBassign\fR commands now in effect.
.SH "CUSTOMIZATION OPTIONS"
.TP
.TP
\fB-a \fIname\fR
Use \fIname\fR as the actual file name for the file specified by
\fIfilespec\fR, overriding any name provided by the \fIFILE=\fR
specifier in an \fIOPEN\fR statement.
.TP
\fB-F \fIconversion\fR Convert the byte ordering of each record header
(a four-byte record length which appears at the beginning and end of
each record) in a sequential unformatted file.  It is allowed only
with unformatted files.  The option has no effect on the data itself
(see the \fB\-N\fR option for that).  The \fIconversion\fR may be:
.RS
.TP 4
\fBf77.mips\fR
On a little-endian computer, use big-endian format for each record header.
.TP
\fBf77.vax\fR
On a big-endian computer, use little-endian format for each record header.
.RE
.TP
\fB\-N \fIconversion\fR
Convert the byte ordering of the data in an unformatted file. It is
allowed only with unformatted files. For sequential unformatted files,
which have record headers, you will usually want to use the \fB\-F\fR
option as well so that the record headers are also converted. The
\fIconversion\fR may be:
.RS
.TP 4
\fBbe\fR
On a little-endian computer, use big-endian format for the data.
.TP
\fBia64\fR
On a big-endian computer, use little-endian format for the data.
.RE
.TP
\fB\-S on|off\fR 
The \fB\-S off\fR option generates commas in list-directed or namelist
output on files specified by \fIfilespec\fR. The default is \fB\-S
on\fR. For example, if the default output would be "4 5 6", \fB\-S off\fR
produces "4, 5, 6".
.TP
\fB\-y on|off\fR
The \fB\-y off\fR option allows the use of repetition factors in
list-directed or namelist
output on files specified by \fIfilespec\fR. The default
is \fB\-y on\fR.  For example, if the default output would be "4 4 4",
\fB\-y off\fR produces "3*4".
.SH "EXAMPLES"
Assign name \fItmp\fR to unit 8, removing any previous \fBassign\fR commands
whether or not they relate to unit 8:
.PP
.ft B
        % assign -a tmp u:8
.ft R
.PP
Assign name \fItmp\fR to unit 8 without removing any previous \fBassign\fR
commands:
.PP
.ft B
        % assign -I -a tmp u:8
.ft R
.PP
Convert numeric data from big-endian format when reading file
\fIdata\fR, and convert numeric data to big-endian format when writing
that file (note that if the file is sequential, its record headers
will remain in little-endian form unless you also use the \fB\-F\fR
option):
.PP
.ft B
        % assign -I -N be f:data
.ft R
.PP
Allow repetition factors in list-directed output on any file whose name
begins with \fBx\fR and ends with \fBy\fR:
.PP
.ft B
        % assign -I -y off p:x%y
.ft R
.PP
For all direct unformatted files, convert numeric data from big-endian
format on input, and generate big-endian format on output:
.PP
.ft B
        % assign -I -N be g:du
.ft R
.PP
For all sequential unformatted files, convert data from big-endian format on
input, and generate big-endian format on output ( the \fB\-N\fR option); do
likewise with the record headers (the \fB\-F f77.mips\fR option):
.PP
.ft B
        % assign -I -N be -F f77.mips g:su
.ft R
.PP
Allow the use of repetition factors on standard output (unit 6):
.PP
.ft B
	% assign -I -y off u:6
.ft R
.SH "USING ASSIGN WITH MPI"
When using assign with MPI, make sure that the environment variable FILENV is
properly defined and that the assign file is accessible by each process. The
mpirun program is not going to set the environment variable for you, so you need
to have a wrapper-script that sets it. 
.PP
You can  define FILENV in your .cshrc or .bashrc file with these commands:
.PP
For bash:
.PP
.ft B
#The following will concatenate the line 'export FILENV=.assign' to the
#end of your .bashrc. The ${HOME}/ is so that you can use this from any
#directory.
  % echo "export FILENV=.assign" >> ${HOME}/.bashrc
.ft R
.PP
For csh:
.PP
.ft B
#The following will concatenate the line 'setenv FILENV .assign' to the
#end of your .cshrc.  The ${HOME}/ is so that you can use this from any
#directory
  % echo "setenv FILENV .assign" >>  ${HOME}/.cshrc
.ft R 
.PP
If you are using an MPI process with a daemon startup, define FILENV
by passing environment variables through a daemon. This is often MPI
implementation-dependent.
.PP
For example, LAM/MPI supports defining FILENV through the \fB\-x\fR option on
mpirun, so your commands would look something like this (in bash):
.PP
.ft B
        % export FILENV=.assign
        % mpirun -np 2 -x FILENV foo
.ft R
.PP
This would launch the \fIfoo\fR program on 2 processes, each aware of the
FILENV variable.
.SH "ENVIRONMENT"
.TP
\fBFILENV\fR
The location of the assign file.
.SH "COPYRIGHT"
Copyright (C) 2007 PathScale, LLC.  All Rights Reserved.
.PP
Copyright (C) 2006, 2007 QLogic Corp.  All Rights Reserved.
.PP
Copyright (C) 2004, 2005 PathScale, Inc.  All Rights Reserved.
.SH "SEE ALSO"
.PP
\fBpathf95\fR(1)
.PP
PathScale Compiler Suite User Guide
.PP
Online documentation available at http://www.pathscale.com/docs.html


pathscale/tools/assign/assign.c
/*
 * Copyright 2004, 2005, 2006 PathScale Inc.  All Rights Reserved.
 */

/*
 
   Path64 is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation version 3

   Path64 is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
   License for more details.

   You should have received a copy of the GNU General Public License
   along with Path64; see the file COPYING.  If not, write to the Free
   Software Foundation, 51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301, USA.

*/


#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <fortran.h>
#include <cray/assign.h>

/*
 * Check if a string is a simple integer - i.e. that all elements in
 * the string are digits.
 */

void
check_int(char *s, char *o, int *error)
{
    int i;

    for(i = 0; i < strlen(s); i++) {
        if(!isdigit(s[i])) {
            fprintf(stderr, "Illegal %s value: %s is not an integer.\n", o, s);
            *error = 1;
            return;
        }
    }
}


void
check_on_off(char *s, char *o, int *error)
{
    if(strcmp(s, "on") && strcmp(s, "off")) {
        fprintf(stderr, "%s is not a valid setting for %s.  "
                "Valid settings are on or off.\n", s, o);
        *error = 1;
    }
}


void
parse_F_spec(char *class, int *error)
{
    char *type = NULL, *subtype = NULL;
    char *num1, *num2 = NULL, *num3 = NULL;
    char *c = strdup(class);

    num1 = index(c, ':');
    if(num1) {
        *num1 = '\0';
        num1 = num1 + 1;
        num2 = index(num1, ':');
        if(num2) {
            *num2 = '\0';
            num2 = num2 + 1;
            num3 = index(num2, ':');
            if(num3) {
                *num3 = '\0';
                num3 = num3 + 1;
            }
        }
    }

    type = index(c, '.');
    if(type) {
        *type = '\0';
        type = type + 1;
        subtype = index(type, '.');
        if(subtype) {
            *subtype = '\0';
            subtype = subtype + 1;
        }
    }

    if(strcmp(c, "blankx") && strcmp(c, "blx") &&
       strcmp(c, "bufa") && strcmp(c, "cache") &&
       strcmp(c, "cachea") && strcmp(c, "cos") &&
       strcmp(c, "blocked") && strcmp(c, "event") &&
       strcmp(c, "f77") && strcmp(c, "fd") &&
       strcmp(c, "global") && strcmp(c, "ibm") &&
       strcmp(c, "mr") && strcmp(c, "null") &&
       strcmp(c, "site") && strcmp(c, "stdin") &&
       strcmp(c, "stdout") && strcmp(c, "stderr") &&
       strcmp(c, "syscall") && strcmp(c, "system") &&
       strcmp(c, "text") && strcmp(c, "user") &&
       strcmp(c, "vms")) {
        fprintf(stderr, "Unknown FFIO class %s.\n", c);
        *error = 1;
    }

    if(num1) check_int(num1, "FFIO class numeric", error);
    if(num2) check_int(num2, "FFIO class numeric", error);
    if(num3) check_int(num3, "FFIO class numeric", error);
}


void
parse_F(char *s, int *error)
{
    char *current = s;

    do {
        char *comma, *nc;
        comma = index(current, ',');
        if(comma) {
            *comma = '\0';
            nc = comma + 1;
        } else {
            nc = &(current[strlen(current)]);
        }
        parse_F_spec(current, error);
        current = nc;
    } while (*current);
}


int
main (int argc, char **argv)
{
    int error = 0;
    int found_a = 0;
    int found_D = 0;
    int found_I = 0;
    int found_O = 0;
    int found_V = 0;
    int found_R = 0;
    int i, l = 0;
    char *str;
    FILE *ef;
    assign_environment aep;

    while(1) {
        int c;

        c = getopt(argc, argv, "a:b:f:m:s:tu:y:B:C:D:F:IN:ORS:T:U:VW:Y:Z:");
        if(c == -1) break;
        switch(c) {
        case 'a':
            found_a = 1;
            break;
        case 'b':
            check_int(optarg, "-b", &error);
            break;
        case 'f':
            if(strcmp(optarg, "77") && strcmp(optarg, "90")) {
                printf("%s is not a valid setting for -f.  Valid settings are "
                       "77 or 90.\n", optarg);
                error = 1;
            }
            break;
        case 'm':
            check_on_off(optarg, "-m", &error);
            break;
        case 's':
            if(strcmp(optarg, "bin") && strcmp(optarg, "cos") &&
               strcmp(optarg, "blocked") && strcmp(optarg, "sbin") &&
               strcmp(optarg, "text") && strcmp(optarg, "u") &&
               strcmp(optarg, "unblocked")) {
                printf("%s is not a valid setting for -s.\nValid settings are "
                       "bin, cos, blocked, sbin, text, u or unblocked.\n",
                       optarg);
                error = 1;
            }
            break;
        case 't':
            break;
        case 'u':
            check_int(optarg, "-u", &error);
            break;
        case 'y':
            check_on_off(optarg, "-y", &error);
            break;
        case 'B':
            check_on_off(optarg, "-B", &error);
            break;
        case 'C':
            if(strcmp(optarg, "ascii") && strcmp(optarg, "(no-op)") &&
               strcmp(optarg, "ebcdic")) {
                fprintf(stderr, "%s is not a valid setting for -C.  "
                        "Valid settings are ascii, (no-op) or ebcdic.\n",
                        optarg);
                error = 1;
            }
            break;
        case 'D':
            if(strcmp(optarg, "stdin") && strcmp(optarg, "stdout") &&
               strcmp(optarg, "stderr")) {
                fprintf(stderr, "%s is not a valid setting for -D.  "
                        "Valid settings are stdin, stdout or stderr.\n",
                        optarg);
                error = 1;
            }
            found_D = 1;
            break;
        case 'F':
            parse_F(optarg, &error);
            break;
        case 'I':
            found_I = 1;
            break;
        case 'N':
            if(strcmp(optarg, "cray") && strcmp(optarg, "ibm") &&
               strcmp(optarg, "vms") && strcmp(optarg, "ieee") &&
               strcmp(optarg, "ieee_32") && strcmp(optarg, "mips") &&
               strcmp(optarg, "ieee_64") && strcmp(optarg, "ieee_le") &&
               strcmp(optarg, "ultrix") && strcmp(optarg, "t3e") &&
               strcmp(optarg, "t3d") && strcmp(optarg, "user") &&
               strcmp(optarg, "site") && strcmp(optarg, "ia32") &&
               strcmp(optarg, "ia64") && strcmp(optarg, "be") &&
               strcmp(optarg, "bigendian")) {
                fprintf(stderr, "%s is not a valid setting for -N.  "
                        "Valid settings are cray, ibm, vms, ieee, ieee_32, "
                        "mips, ieee_64, ieee_le, ultrix, t3e, t3d, user, "
                        "site, ia32, ia64, be or bigendian.", optarg);
            }
            break;
        case 'O':
            found_O = 1;
            break;
        case 'R':
            found_R = 1;
            break;
        case 'S':
            check_on_off(optarg, "-S", &error);
            break;
        case 'T':
            check_on_off(optarg, "-T", &error);
            break;
        case 'U':
            check_on_off(optarg, "-U", &error);
            break;
        case 'V':
            found_V = 1;
            break;
        case 'W':
            check_on_off(optarg, "-W", &error);
            break;
        case 'Y':
            check_on_off(optarg, "-Y", &error);
            break;
        case 'Z':
            check_on_off(optarg, "-Z", &error);
            break;
        default:
            error = 1;
            break;
        }
    }

    if(found_a && found_D) {
        fprintf(stderr, "-a and -D options are mutually exclusive.");
        error = 1;
    }

    if(found_I && found_O) {
        fprintf(stderr, "-I and -O options are mutually exclusive.");
        error = 1;
    }

    if(optind != argc - 1 && !found_V && !found_R) {
        fprintf(stderr, "usage: %s options assign_object\n", argv[0]);
        error = 1;
    }

    /*
     * If we do "assign -R something" then add a -I to the mix.
     */

    if(optind == argc - 1 && found_R) {
        found_I = 1;
    }

    if(optind != argc) {
        if(!strncmp(argv[optind], "g:", 2)) {
            if(strcmp(argv[optind], "g:all") && strcmp(argv[optind], "g:su") &&
               strcmp(argv[optind], "g:sf") && strcmp(argv[optind], "g:du") &&
               strcmp(argv[optind], "g:df") && strcmp(argv[optind], "g:aq") &&
               strcmp(argv[optind], "g:ff")) {
                fprintf(stderr, "Illegal g: assign object %s.  Valid g: "
                        "options are g:all, g:su, g:sf, g:du, g:df, g:aq or "
                        "g:ff.\n", argv[optind]);
            error = 1;
            }
        } else if(!strncmp(argv[optind], "u:", 2)) {
            check_int(&(argv[optind][2]), "u:", &error);
        }
    }

    if(!getenv("FILENV")) {
        fprintf(stderr, "FILENV environment variable not set.\n");
        error = 1;
    }

    if(error) {
        return 1;
    }

    _ae_setupenv(&aep);
    _Ae_env_stack.env = &aep;
    _Ae_env_stack.size++;

    if(found_I || found_V) {
        ef = _gae_open('r', 'u', &error);
        if(ef == NULL && error) {
            fprintf(stderr, "Couldn't read assign file: %s.\n",\
                    strerror(errno));
            return 1;
        }

        if(ef) {
            _ae_internalize(ef, &aep);
            _gae_close(ef);
        }
    }

    if(found_V) {
        _ae_externalize(AE_ASGCMD, stdout, &aep);
        return 0;
    }

    for(i = 1; i < argc; i++)
        l += strlen(argv[i]) + 1;

    str = (char *)malloc(l + strlen("assign "));
    strcpy(str, "assign ");

    for(i = 1; i < argc; i++) {
        strcat(str, argv[i]);
        if(i != argc - 1) {
            strcat(str, " ");
        }
    }

    l = strlen(str);

    error = ffassign(str);

    ef = _gae_open('w', 'u', &error);
    if(ef == NULL) {
        fprintf(stderr, "Couldn't write assign file: %s.\n", strerror(errno));
        return 1;
    }

    _ae_externalize(AE_ASGCMD, ef, &aep);

    _gae_close(ef);

    return 0;
}
